File: src/main/java/dev/harrel/jarhell/model/Gav.java
package dev.harrel.jarhell.model;

import java.util.Objects;
import java.util.StringJoiner;

public record Gav(String groupId, String artifactId, String version, String classifier) {
    public Gav {
        Objects.requireNonNull(groupId, "Field 'groupId' is required");
        Objects.requireNonNull(artifactId, "Field 'artifactId' is required");
        Objects.requireNonNull(version, "Field 'version' is required");
    }

    public Gav(String groupId, String artifactId, String version) {
        this(groupId, artifactId, version, null);
    }

    public static Gav fromCoordinate(String coordinate) {
        String[] split = coordinate.split(":");
        if (split.length == 3) {
            return new Gav(split[0], split[1], split[2]);
        } else if (split.length == 4) {
            return new Gav(split[0], split[1], split[2], split[3]);
        } else {
            throw new IllegalArgumentException("Invalid coordinate format");
        }
    }

    public Gav stripClassifier() {
        return classifier == null ? this : new Gav(groupId, artifactId, version);
    }

    @Override
    public String toString() {
        StringJoiner joiner = new StringJoiner(":")
                .add(groupId)
                .add(artifactId)
                .add(version);
        if (classifier != null) {
            joiner.add(classifier);
        }
        return joiner.toString();
    }
}

File: web/src/utils/gav.ts

export interface Gav {
  groupId: string
  artifactId: string
  version?: string | undefined
}

export function parseGav(data: string | undefined): Gav | null {
  if (!data) {
    return null
  }
  const parts = data.split(':')
  if (parts.length !== 2 && parts.length !== 3) {
    return null
  }
  return {
    groupId: parts[0]!,
    artifactId: parts[1]!,
    version: parts[2]
  }
}